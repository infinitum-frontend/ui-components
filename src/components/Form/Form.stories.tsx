// eslint-disable-next-line @typescript-eslint/no-unused-vars
import React, { FormEventHandler, useState } from 'react'
import { StoryFn, Meta } from '@storybook/react'
import { Form } from './index'
import { Button } from '../Button'
import { Checkbox } from '../Checkbox'
import { Input } from '../Input'
import { Select } from '../Select'
import { Radio, RadioGroup } from '../Radio'
import { Textarea } from '../Textarea'
import { Switch } from '../Switch'

const meta: Meta<typeof Form> = {
  title: 'Form/Form',
  component: Form,
  subcomponents: {
    'Form.Group': Form.Group,
    'Form.Label': Form.Label,
    'Form.Hint': Form.Hint,
    'Form.Item': Form.Item,
    'Form.ErrorMessage': Form.ErrorMessage
  }
}

export default meta

const Template: StoryFn<typeof Form> = (args, subArgs) => {
  const [email, setEmail] = useState('')
  const [emailError, setEmailError] = useState<boolean>(true)

  const handleInput = (value: string): void => {
    setEmail(value)
    setEmailError(false)
  }

  const handleSubmit: FormEventHandler = (e): void => {
    if (!email.length) {
      setEmailError(true)
    } else {
      alert('submit')
    }
  }

  return (
    <Form {...args} onSubmit={handleSubmit} labelWidth={'150px'}>
      <Form.Group gap={'xsmall'} isRequired={true}>
        <Form.Label>Email</Form.Label>
        <Form.Item>
          <Input
            value={email}
            onInput={handleInput}
            status={emailError ? 'error' : undefined}
          />
        </Form.Item>

        {emailError ? (
          <Form.ErrorMessage>
            Введен невалидный почтовый адрес
          </Form.ErrorMessage>
        ) : (
          <Form.Hint>Обязательное поле</Form.Hint>
        )}
      </Form.Group>
      <Button type={'submit'}>Отправить</Button>
    </Form>
  )
}

export const Playground = Template.bind({})

export const WithNativeValidation: StoryFn<typeof Form> = (args) => {
  const [comment, setComment] = useState<string>('')
  const [restrictionType, setRestrictionType] = useState<string>('')
  const [fullName, setFullName] = useState<string>('')
  const [entityType, setEntityType] = useState<string>('')
  const [subscription, setSubscription] = useState<boolean>(false)

  const handleSubmit: FormEventHandler = (e): void => {
    alert('submit')
  }

  return (
    <>
      <code style={{ display: 'block', marginBottom: '12px' }}>
        Есть возможность использовать нативную валидацию. Работает для
        компонентов Input, Textarea, Switch, Select, Radio(для RadioGroup есть
        также проп required, обрабатывающий вложенные радиокнопки), Checkbox
      </code>
      <Form {...args} onSubmit={handleSubmit} labelWidth={'200px'}>
        <Form.Group direction={'horizontal'}>
          <Form.Label>ФИО</Form.Label>
          <Form.Item>
            <Input
              required={true}
              value={fullName}
              onInput={(value) => setFullName(value)}
              minLength={6}
            />
          </Form.Item>
        </Form.Group>

        <Form.Group direction={'horizontal'}>
          <Form.Label>Вид сущности</Form.Label>
          <Form.Item>
            <Select
              required={true}
              value={entityType}
              onChange={(option) => setEntityType(option.value as string)}
              options={[
                { label: 'Акции', value: '1' },
                { label: 'Облигации', value: '2' }
              ]}
            />
          </Form.Item>
        </Form.Group>

        <Form.Group direction={'horizontal'}>
          <Form.Label>Комментарий в ТЗ</Form.Label>
          <Form.Item>
            <Textarea
              resize={'vertical'}
              required={true}
              maxLength={10}
              value={comment}
              onInput={(value) => setComment(value)}
            />
          </Form.Item>
          <Form.Hint>Комментарий виден только сотрудникам</Form.Hint>
        </Form.Group>

        <Form.Group direction={'horizontal'}>
          <Form.Label>Счёт депо</Form.Label>
          <Form.Item>
            <Checkbox required={true}>
              Использовать для отбора счёт депо
            </Checkbox>
          </Form.Item>
        </Form.Group>

        <Form.Group direction={'horizontal'}>
          <Form.Label>Ограничение в обороте</Form.Label>
          <Form.Item>
            <RadioGroup
              name={'restrictions'}
              value={restrictionType}
              onChange={(value) => setRestrictionType(value)}
              required={true}
            >
              <Radio value={'1'}>Ограничено</Radio>
              <Radio value={'2'}>Не ограничено</Radio>
              <Radio value={'3'}>Не указано</Radio>
            </RadioGroup>
          </Form.Item>
        </Form.Group>

        <Form.Group direction={'horizontal'}>
          <Form.Label>Получать уведомления на почту</Form.Label>
          <Form.Item>
            <Switch
              checked={subscription}
              onChange={(checked) => setSubscription(Boolean(checked))}
            />
          </Form.Item>
        </Form.Group>

        <Button type={'submit'}>Подтвердить</Button>
      </Form>
    </>
  )
}
