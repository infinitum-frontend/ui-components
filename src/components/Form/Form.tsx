// eslint-disable-next-line @typescript-eslint/no-unused-vars
import React, {
  ComponentPropsWithRef,
  CSSProperties,
  FormEventHandler,
  forwardRef,
  ReactElement,
  useRef
} from 'react'
import cn from 'classnames'
import { Space, SpaceProps } from 'Components/Space'
import FormContext from './context/form'
import FormLabel from './components/FormLabel'
import FormGroup from './components/FormGroup'
import FormHint from './components/FormHint'
import FormItem from './components/FormItem'
import FormErrorMessage from './components/FormErrorMessage'
import FormAction from './components/FormAction'
import { mergeRefs } from 'react-merge-refs'

export interface FormProps extends ComponentPropsWithRef<'form'> {
  /** Расстояние между блоками */
  gap?: SpaceProps['gap']
  /** Состояние недоступности */
  disabled?: boolean
  /** Ширина лейбла (валидное значение css-width) */
  labelWidth?: CSSProperties['width']
}

const Form = forwardRef<HTMLFormElement, FormProps>(
  (
    {
      gap = 'medium',
      children,
      labelWidth = '',
      onSubmit,
      className,
      disabled,
      ...props
    },
    ref
  ): ReactElement => {
    const innerRef = useRef<HTMLFormElement>(null)

    const handleSubmit: FormEventHandler<HTMLFormElement> = (e) => {
      e.preventDefault()

      const form = e.target as HTMLFormElement
      const isValid = form.checkValidity()
      if (!isValid) {
        return false
      }

      onSubmit?.(e)
    }

    const context = {
      labelWidth,
      form: innerRef,
      disabled
    }

    return (
      <FormContext.Provider value={context}>
        <Space
          as={'form'}
          gap={gap}
          onSubmit={handleSubmit}
          noValidate={true}
          ref={mergeRefs([innerRef, ref])}
          className={cn('inf-form', className, {})}
          {...props}
        >
          {children}
        </Space>
      </FormContext.Provider>
    )
  }
)

Form.displayName = 'Form'

/** Компонент для создания интерактивных и удобных форм ввода, упрощающий пользователям ввод и отправку данных с проверками и настраиваемыми полями */
export default Object.assign(Form, {
  Group: FormGroup,
  Item: FormItem,
  Label: FormLabel,
  Hint: FormHint,
  ErrorMessage: FormErrorMessage,
  Action: FormAction
})
