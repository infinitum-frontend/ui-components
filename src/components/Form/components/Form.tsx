// eslint-disable-next-line @typescript-eslint/no-unused-vars
import React, {
  ComponentPropsWithRef,
  CSSProperties,
  FormEventHandler,
  forwardRef,
  ReactElement
} from 'react'
import cn from 'classnames'
import FormLabel from 'Components/Form/components/Label'
import FormGroup from 'Components/Form/components/Group'
import FormHint from 'Components/Form/components/Hint'
import Space, { SpaceProps } from 'Components/Space/Space'
import FormContext from 'Components/Form/context/form'
import FormItem from 'Components/Form/components/Item'
import FormErrorMessage from 'Components/Form/components/ErrorMessage'

export interface FormProps extends ComponentPropsWithRef<'form'> {
  /**
   * Расстояние между блоками
   */
  gap?: SpaceProps['gap']
  /** Ширина лейбла (валидное значение css-width) */
  labelWidth?: CSSProperties['width']
}

const Form = forwardRef<HTMLFormElement, FormProps>(
  (
    {
      gap = 'medium',
      children,
      labelWidth = '',
      onSubmit,
      className,
      ...props
    },
    ref
  ): ReactElement => {
    const handleSubmit: FormEventHandler<HTMLFormElement> = (e) => {
      e.preventDefault()
      onSubmit?.(e)
    }

    const context = {
      labelWidth
    }

    return (
      <FormContext.Provider value={context}>
        <Space
          as={'form'}
          gap={gap}
          onSubmit={handleSubmit}
          ref={ref}
          className={cn('inf-form', className, {})}
          {...props}
        >
          {children}
        </Space>
      </FormContext.Provider>
    )
  }
)

Form.displayName = 'Form'

/** Компонент для создания интерактивных и удобных форм ввода, упрощающий пользователям ввод и отправку данных с проверками и настраиваемыми полями */
export default Object.assign(Form, {
  Group: FormGroup,
  Item: FormItem,
  Label: FormLabel,
  Hint: FormHint,
  ErrorMessage: FormErrorMessage
})
