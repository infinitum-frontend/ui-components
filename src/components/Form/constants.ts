export const formElementDisplayName = {
  Group: 'Form.Group',
  Item: 'Form.Item',
  Label: 'Form.Label',
  Hint: 'Form.Hint',
  ErrorMessage: 'Form.ErrorMessage',
  Action: 'Form.Action'
}
