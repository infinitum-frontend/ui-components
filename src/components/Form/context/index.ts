import { createContext, useContext } from 'react'

export interface IFormContext {
  id?: string
}

const FormContext = createContext<IFormContext | null>(null)

export function useForm(): IFormContext | null {
  return useContext(FormContext)
}

export default FormContext
