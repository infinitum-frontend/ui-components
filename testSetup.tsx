import { expect, beforeAll } from 'vitest'
import { render, RenderOptions } from '@testing-library/react'
// import '@testing-library/jest-dom/vitest' for vitest 1.x.x
import { ReactElement, JSXElementConstructor } from 'react'
import matchers from '@testing-library/jest-dom/matchers'

// extends Vitest's expect method with methods from react-testing-library
// https://github.com/testing-library/jest-dom#custom-matchers
expect.extend(matchers)

export const renderComponent = function (
  Component: ReactElement,
  options?: RenderOptions
): {
  el: HTMLElement
  container: HTMLElement
  rerender: (ui: ReactElement<any, string | JSXElementConstructor<any>>) => void
} {
  const { container, rerender } = render(Component, options)

  const el = container.firstChild as HTMLElement

  return { el, container, rerender }
}

beforeAll(() => {
  global.ResizeObserver = class ResizeObserver {
    observe(): any {}
    unobserve(): any {}
    disconnect(): any {}
  }

  // https://github.com/testing-library/react-testing-library/issues/95#issuecomment-392879866
  // https://github.com/vuejs/vue-test-utils/issues/319
  if (!HTMLElement.prototype.scrollTo) {
    HTMLElement.prototype.scrollTo = () => {}
  }
})

// runs a cleanup after each test case (e.g. clearing jsdom)
// afterEach(() => {
//   cleanup()
// })
