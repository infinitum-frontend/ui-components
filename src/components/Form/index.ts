import Form, { FormProps } from './Form'
import { FormGroupProps } from './components/FormGroup'

export { Form }
export type { FormProps, FormGroupProps }
