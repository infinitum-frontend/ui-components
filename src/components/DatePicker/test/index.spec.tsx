// Тест кейсы:
// 1. компонент рендерится корректно +
// 2. поддерживает пропс disabled +
// 3. поддерживает пропс placeholder +
// 4. поддерживает пропс size +
// 5. поддерживает пропс value +
// 6. поддерживает пропс min/max +
// 7. поддерживает пропс clearable +
// 8. поддерживает пропс withTodayButton +
// 9. вызывает onChange при выборе даты +
// 10. открывает/закрывает календарь при клике +
// 11. поддерживает пропс required +
// 12. работает с Form контекстом +
// 13. поддерживает пропс onClear +

import { renderComponent } from '~/testSetup'
import { DatePicker } from '../index'
import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Form } from '../../Form'
import { vi } from 'vitest'

describe('DatePicker', () => {
  it('should match snapshot', () => {
    const { el } = renderComponent(<DatePicker />)
    expect(el).toMatchSnapshot()
  })

  it('should render correctly', () => {
    renderComponent(<DatePicker />)

    const input = screen.queryByRole('textbox')
    expect(input).toBeInTheDocument()
    expect(input).toHaveAttribute('placeholder', '__.__.____')
  })

  it('should support disabled prop', () => {
    renderComponent(<DatePicker disabled />)

    const input = screen.queryByRole('textbox')
    expect(input).toBeDisabled()
  })

  it('should support custom placeholder', () => {
    const customPlaceholder = 'Выберите дату'
    renderComponent(<DatePicker placeholder={customPlaceholder} />)

    const input = screen.queryByRole('textbox')
    expect(input).toHaveAttribute('placeholder', customPlaceholder)
  })

  it('should support size prop', () => {
    const { el } = renderComponent(<DatePicker size="small" />)

    const input = el.querySelector('.inf-input-wrapper')
    expect(input).toHaveClass('inf-text-field--size-small')
  })

  it('should support value prop', () => {
    const testDate = '2024-01-15'
    renderComponent(<DatePicker value={testDate} />)

    const input = screen.queryByRole('textbox')
    expect(input).toHaveValue('15.01.2024')
  })

  it('should support min/max props', () => {
    const minDate = '2024-01-01'
    const maxDate = '2024-12-31'
    renderComponent(<DatePicker min={minDate} max={maxDate} />)

    const input = screen.queryByRole('textbox')
    expect(input).toHaveAttribute('min', minDate)
    expect(input).toHaveAttribute('max', maxDate)
  })

  it('should support clearable prop', async () => {
    const { el } = renderComponent(<DatePicker clearable value="2024-01-15" />)

    const clearButton = el.querySelector(
      '.inf-input-wrapper__clear-button'
    ) as HTMLSpanElement

    expect(clearButton).toBeInTheDocument()
    // expect(clearButton).toBeVisible() не работает из за jsdom
  })

  it('should show clear button only when value exists and clearable is true', () => {
    const { rerender, el } = renderComponent(<DatePicker clearable />)

    // Без значения кнопка очистки не должна отображаться
    expect(
      el.querySelector('.inf-input-wrapper__clear-button') as HTMLSpanElement
    ).not.toBeInTheDocument()

    // С значением кнопка должна отображаться
    rerender(<DatePicker clearable value="2024-01-15" />)
    expect(
      el.querySelector('.inf-input-wrapper__clear-button') as HTMLSpanElement
    ).toBeInTheDocument()
  })

  it('should support withTodayButton prop', async () => {
    renderComponent(<DatePicker withTodayButton />)

    const user = userEvent.setup()
    const input = screen.queryByRole('textbox') as HTMLInputElement

    // Открываем календарь
    await user.click(input)

    // Проверяем наличие кнопки "Сегодня"
    expect(screen.queryByText('Сегодня')).toBeInTheDocument()
  })

  it('should call onChange when date is selected', async () => {
    const onChange = vi.fn()
    renderComponent(<DatePicker onChange={onChange} />)

    const user = userEvent.setup()
    const input = screen.queryByRole('textbox') as HTMLInputElement

    // Открываем календарь
    await user.click(input)

    // Выбираем дату (кликаем по дню)
    const dayButton = screen.queryByText('15')
    if (dayButton) {
      await user.click(dayButton)

      expect(onChange).toHaveBeenCalledWith(
        new Date(new Date().setDate(15)).toISOString().split('T')[0]
      )
    }
  })

  it('should open/close calendar on click', async () => {
    renderComponent(<DatePicker value={'2025-08-15'} />)

    const user = userEvent.setup()
    const input = screen.queryByRole('textbox') as HTMLInputElement

    // Изначально календарь закрыт
    expect(screen.queryByText('Август')).toBeNull()

    // Открываем календарь
    await user.click(input)
    expect(screen.queryByText('Август')).toBeInTheDocument()

    // // // Закрываем календарь (кликаем снова)
    await user.click(document.body)
    expect(screen.queryByText('Август')).toBeNull()
  })

  it('should support required prop', () => {
    renderComponent(<DatePicker required />)

    const input = screen.queryByRole('textbox')
    expect(input).toHaveAttribute('required')
  })

  it('should work with Form context', () => {
    renderComponent(
      <Form>
        <Form.Group required>
          <DatePicker />
        </Form.Group>
      </Form>
    )

    const input = screen.queryByRole('textbox')
    expect(input).toHaveAttribute('required')
  })

  it('should call onClear when clear button is clicked', async () => {
    const onClear = vi.fn()
    renderComponent(
      <DatePicker clearable value="2024-01-15" onClear={onClear} />
    )

    const user = userEvent.setup()
    const clearButton = screen.queryByRole('button', { name: /очистить/i })

    if (clearButton) {
      await user.click(clearButton)
      expect(onClear).toHaveBeenCalledTimes(1)
    }
  })

  it('should call onChange when clear button is clicked without onClear', async () => {
    const onChange = vi.fn()
    renderComponent(
      <DatePicker clearable value="2024-01-15" onChange={onChange} />
    )

    const user = userEvent.setup()
    const clearButton = screen.queryByRole('button', { name: /очистить/i })

    if (clearButton) {
      await user.click(clearButton)
      expect(onChange).toHaveBeenCalledWith('')
    }
  })

  it('should close calendar after date selection', async () => {
    renderComponent(<DatePicker value={'2025-08-15'} />)

    const user = userEvent.setup()
    const input = screen.queryByRole('textbox') as HTMLInputElement

    // Открываем календарь
    await user.click(input)
    expect(screen.queryByText('Август')).toBeInTheDocument()

    // Выбираем дату
    const dayButton = screen.queryByText('15')
    if (dayButton) {
      await user.click(dayButton)

      // Календарь должен закрыться
      expect(screen.queryByText('Август')).not.toBeInTheDocument()
    }
  })

  it('should handle empty value correctly', () => {
    renderComponent(<DatePicker value="" />)

    const input = screen.queryByRole('textbox')
    expect(input).toHaveValue('')
  })

  it('should handle Date object value', () => {
    const testDate = new Date('2024-01-15')
    renderComponent(<DatePicker value={testDate} />)

    const input = screen.queryByRole('textbox')
    expect(input).toHaveValue('15.01.2024')
  })
})
