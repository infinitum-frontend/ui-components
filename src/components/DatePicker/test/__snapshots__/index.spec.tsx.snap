// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NativeDatePicker > should match snapshop 1`] = `
<input
  class="inf-datepicker inf-text-field inf-text-field--br-regular"
  role="calendar"
  type="date"
  value=""
/>
`;
